Python.Python version 3.11.0

remyjette opened the pull request:
- [X] Have you signed the [Contributor License Agreement](https://cla.opensource.microsoft.com/microsoft/winget-pkgs)?
- [X] Have you checked that there aren't other open [pull requests](https://github.com/microsoft/winget-pkgs/pulls) for the same manifest update/change?
- [X] Have you [validated](https://github.com/microsoft/winget-pkgs/blob/master/AUTHORING_MANIFESTS.md#validation) your manifest locally with `winget validate --manifest <path>`?
- [X] Have you tested your manifest locally with `winget install --manifest <path>`?
- [X] Does your manifest conform to the [1.2 schema](https://github.com/microsoft/winget-pkgs/tree/master/doc/manifest/schema/1.2.0)?

Note: `<path>` is the name of the directory containing the manifest you're submitting.

-----


###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/microsoft/winget-pkgs/pull/86190)

remyjette committed:
Python.Python version 3.11.0

remyjette committed:
Split changes to Python 3.10's moniker out to a separate PR

SpecterShell commented:
ARM64 version
https://www.python.org/ftp/python/3.11.0/python-3.11.0-arm64.exe

fernandog commented on manifests/p/Python/Python/3/11/3.11.0/Python.Python.3.11.installer.yaml:
```diff
@@ -0,0 +1,73 @@
+- interactive
+- silent
+- silentWithProgress
+UpgradeBehavior: install
```
shouldn't this setting be `UpgradeBehavior: uninstallPrevious` ?

remyjette replied:
My preference would be for all Python versions to have consistent behavior. I don't have an opinion on the correct setting for UpgradeBehavior, but if it were to change then it should change for all previous versions as well, so I decided to keep this to just adding support for the new version.

fernandog replied:
@denelon I'm new to this. what your thought on this?

remyjette replied:
Given this discussion I think it makes sense to switch to `uninstallPrevious` for Python.

Prior to each major.minor being split into a separate package this would not have been correct as an upgrade from 3.10 to 3.11 (for example) could break code.

However, patch releases are compatible with each other and just contain bugfixes, and more importantly, security fixes. So I don't think it makes any sense to leave the old version around.

This would also match behavior with other package managers and repositories, where you can specifically install a major.minor version of Python, but the patch version will always be latest.

I'll update this PR to switch to uninstallPrevious.

@denelon I can make a PR to change previous versions of Python to match. Can this be done in one PR, or do I need to make a separate PR for each version?

denelon replied:
@remyjette unfortunately, it requires one PR per manifest (package/version). The code on the back end to run validation can't handle figuring out which one failed (or which of multiple failures). We're looking at options to make this better/easier to modify, but it's a messy problem.

remyjette replied:
Okay, I'll make a bunch of PRs then :)

remyjette replied:
Submitted PRs for every version that had `UpgradeBehavior: install` to switch to `UpgradeBehavior: uninstallPrevious` (except the .0 versions because that seemed like unnecessary churn)

SpecterShell replied:
I don't think switching to uninstallprevious is necessary because for minor version upgrade the installer can override the existing installation instead of adding another version.
![image](https://user-images.githubusercontent.com/56779163/198178679-c24d5d65-af20-4029-8150-1c5a944e2e2e.png)

fernandog replied:
My 3.11 was installed side-by-side the 3.10

SpecterShell replied:
Oh it is the patch version (3.10.6 → 3.10.8) that can upgrade directly. Minor versions (3.10 and 3.11) can be installed side-by-side. Sorry for mistakes.
All Python versions in WinGet are categorized by minor versions (`Python.Python.3.10`, `Python.Python.3.9`, etc.) and each minor version can only be upgraded by patch version and doesn't affect each other.

remyjette committed:
Change UpgradeBehavior to uninstallPrevious

remyjette committed:
Fix ProductCode to match GUID of uninstaller dir in Package Cache

My initial PR used the GUID for the Python 3.11.0 Core Interpreter, but
that is just one component of the total Python install, and just
uninstalling that will not remove Python but instead leave it in a bad
state. Updated to use the guid of the dir the uninstaller is stored in
inside `%LOCALAPPDATA%\Package Cache` instead.

remyjette committed:
Add the ARM64 installer too

remyjette commented:
> ARM64 version https://www.python.org/ftp/python/3.11.0/python-3.11.0-arm64.exe

Thanks for pointing that 3.11.0 begins native support for ARM64. Took me a bit to get my ARM64 machine working again as it's been a bit since I last used it, but I was able to add arm64 to this manifest and validate that it works.
